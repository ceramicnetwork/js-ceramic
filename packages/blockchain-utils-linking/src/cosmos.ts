import { AccountId } from 'caip'
import { AuthProvider } from './auth-provider.js'
import { getConsentMessage, LinkProof, asOldCaipString } from './util.js'
import type { Tx, SignMeta } from '@tendermint/sig'
import { hash } from '@stablelib/sha256'
import * as uint8arrays from 'uint8arrays'

const stringEncode = (str: string): string =>
  uint8arrays.toString(uint8arrays.fromString(str), 'base64pad')

// return data in the cosmos unsigned transaction format
export function asTransaction(address: string, message: string): Tx {
  return {
    fee: {
      amount: [{ amount: '0', denom: '' }],
      gas: '0',
    },
    memo: message,
    msg: [
      {
        type: 'cosmos-sdk/MsgSend',
        value: {
          from_address: address,
          to_address: address,
          amount: [{ amount: '0', denom: '0' }],
        },
      },
    ],
  }
}

// generate metadata for signing the transaction
export function getMetaData(): SignMeta {
  return {
    account_number: '1',
    chain_id: 'cosmos',
    sequence: '0',
  }
}

export class CosmosAuthProvider implements AuthProvider {
  readonly isAuthProvider = true

  constructor(
    private readonly provider: any,
    private readonly address: string,
    private readonly chainRef: string
  ) {
    console.warn(
      'WARN: CosmosAuthProvider is not fully supported. You may encounter issues using this.'
    )
  }

  async authenticate(message: string): Promise<string> {
    const accountID = await this.accountId()
    const encodedMsg = stringEncode(message)
    const res = await this.provider.sign(
      asTransaction(accountID.address, encodedMsg),
      getMetaData()
    )
    const digest = hash(uint8arrays.fromString(JSON.stringify(res.signatures[0])))
    return `0x${uint8arrays.toString(digest, 'base16')}`
  }

  async createLink(did: string): Promise<LinkProof> {
    const { message, timestamp } = getConsentMessage(did)
    const accountID = await this.accountId()
    const encodedMsg = stringEncode(message)
    const res = await this.provider.sign(
      asTransaction(accountID.address, encodedMsg),
      getMetaData()
    )
    const signature = stringEncode(JSON.stringify(res.signatures[0]))
    return {
      version: 2,
      message,
      signature,
      account: asOldCaipString(accountID),
      timestamp,
    }
  }

  async accountId(): Promise<AccountId> {
    return new AccountId({
      address: this.address,
      chainId: `cosmos:${this.chainRef}`,
    })
  }

  withAddress(address: string): AuthProvider {
    return new CosmosAuthProvider(this.provider, address, this.chainRef)
  }
}
