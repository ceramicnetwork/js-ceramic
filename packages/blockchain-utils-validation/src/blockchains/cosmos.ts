import { AccountID } from "caip";
import { verifyTx } from '@tendermint/sig';
import { BlockchainHandler } from "../blockchain-handler";
import { LinkProof, cosmos } from "@ceramicnetwork/blockchain-utils-linking";
import * as uint8arrays from "uint8arrays";

const namespace = "cosmos";

const stringEncode = (str: string): string => uint8arrays.toString(uint8arrays.fromString(str), 'base64pad');
const stringDecode = (str: string): string => uint8arrays.toString(uint8arrays.fromString(str, 'base64pad'));

export async function validateLink(
  proof: LinkProof
): Promise<LinkProof | null> {
    const account = new AccountID(proof.account);
    const encodedMsg = stringEncode(proof.message);
    const payload = cosmos.asTransaction(account.address, encodedMsg);
    const sigObj = JSON.parse(stringDecode(proof.signature));
    const Tx = { ...payload, ...cosmos.getMetaData(), signatures: [sigObj] };
    const is_sig_valid = verifyTx(Tx, cosmos.getMetaData());
    return is_sig_valid ? proof : null;
}

const Handler: BlockchainHandler = {
  namespace,
  validateLink,
};

export default Handler;
