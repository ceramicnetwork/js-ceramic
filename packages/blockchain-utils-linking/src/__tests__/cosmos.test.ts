import { jest } from '@jest/globals'
// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore
import tendermint from '@tendermint/sig'
import { CosmosAuthProvider } from '../cosmos.js'

const did = 'did:3:bafysdfwefwe'
const mnemonic = 'test salon husband push melody usage fine ensure blade deal miss twin'
const local_provider = tendermint.createWalletFromMnemonic(mnemonic)
const chainRef = 'cosmoshub-3'

class CosmosMockSigner {
  readonly provider: tendermint.Wallet

  constructor(local_provider: tendermint.Wallet) {
    this.provider = local_provider
  }

  public async sign(msg: tendermint.Tx, metadata: tendermint.SignMeta): Promise<tendermint.StdTx> {
    return new Promise((resolve): void => {
      const signature = tendermint.signTx(msg, metadata, this.provider)
      resolve(signature)
    })
  }
}

beforeAll(() => {
  global.Date.now = jest.fn().mockImplementation(() => 666000)
})

afterAll(() => {
  jest.clearAllMocks()
})

describe('Blockchain: Cosmos', () => {
  describe('createLink', () => {
    test('create proof for cosmoshub3', async () => {
      const provider = new CosmosMockSigner(local_provider)
      const authProvider = new CosmosAuthProvider(provider, local_provider.address, chainRef)
      const proof = await authProvider.createLink(did)
      expect(proof).toMatchSnapshot()
    })
  })

  describe('authenticate', () => {
    test('create proof for cosmoshub3', async () => {
      const provider = new CosmosMockSigner(local_provider)
      const authProvider = new CosmosAuthProvider(provider, local_provider.address, chainRef)
      const result = await authProvider.authenticate('msg')
      expect(result).toMatchSnapshot()
    })
  })
})
