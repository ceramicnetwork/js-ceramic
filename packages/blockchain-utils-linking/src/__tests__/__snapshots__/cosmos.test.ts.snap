// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Blockchain: Cosmos authenticate create proof for cosmoshub3 1`] = `"0x3e779670d353ea4826096f99ba6ba6a23ac4efb67de9cffb03909e6017245965"`;

exports[`Blockchain: Cosmos createLink create proof for cosmoshub3 1`] = `
Object {
  "account": "cosmos1gs6gu0w9xvqcfhm3nlhnmqkf8aphnsxyjlgpqd@cosmos:cosmoshub-3",
  "message": "Link this account to your identity

did:3:bafysdfwefwe 
Timestamp: 666",
  "signature": "eyJzaWduYXR1cmUiOiJWbm5IOGpPMXhoU0VNOEZXYTJxTU1QMSs1OXVCdm1WYitzNUwwMjA1TVZONzdHc3pkTzM0T1JPZGpySmNkSSsvMjJPbWt2VkRpRWYyZ2lNZHdRYXJIUT09IiwicHViX2tleSI6eyJ0eXBlIjoidGVuZGVybWludC9QdWJLZXlTZWNwMjU2azEiLCJ2YWx1ZSI6IkE0MVhFTUx0TnhvbmhlU2FwNTg5Nzdkc0ZhSTh5VWt3NHdWSytiR3dPNElyIn19",
  "timestamp": 666,
  "version": 2,
}
`;
